lua/lspconfig/rust_analyzer.lua

```diff
@@ -16,15 +16,27 @@ configs.rust_analyzer = {
     cmd = {"rust-analyzer"};
     filetypes = {"rust"};
     root_dir = function(fname)
-      local cargo_metadata = vim.fn.system("cargo metadata --no-deps --format-version 1")
-      local cargo_root = nil
+      -- Order of preference for how we choose the root dir:
+      --   * Current Cargo workspace (if any)
+      --   * Current Cargo crate (if any)
+      --   * Rust project root, for projects that don't use Cargo (if any)
+      --   * Current git repository
+      local cargo_crate_dir = util.root_pattern("Cargo.toml")(fname)
+      -- Make sure that we run `cargo metadata` in the current project dir
+      -- rather that the dir from which nvim was initially launched.
+      local cmd = "cargo metadata --no-deps --format-version 1"
+      if cargo_crate_dir ~= nil then
```

Comment:
Does this still work correctly in multi-crate projects?

Reply:
If a project has multiple crates that depend on each other they will have to specify the dependency by path (and potentially also belong to the same workspace). I just tested and starting rust-analyzer with a `root_dir` set to an individual crate in such a multi-crate project correctly resolves the dependencies and I can `vim.lsp.buf.definition()` across them.